kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt

```diff
@@ -1,72 +1,1061 @@
+                              { this shouldBeExactly other },
+                              { this shouldBe exactly(other) }
+    )
+  }
+  
+  private fun Double.shouldMatchBetween(a: Double, b: Double, tolerance: Double) {
+    this.shouldBeBetween(a, b, tolerance)
+    this shouldBe between(a, b, tolerance)
+    
+    this.shouldThrowExceptionOnNotBetween(a, b, tolerance)
+  }
+  
+  private fun Double.shouldNotMatchBetween(a: Double, b: Double, tolerance: Double) {
+    this.shouldNotBeBetween(a, b, tolerance)
+    this shouldNotBe between(a, b, tolerance)
+    
+    this.shouldThrowExceptionOnBetween(a, b, tolerance)
+  }
+  
+  fun Double.shouldThrowExceptionOnBetween(a: Double, b: Double, tolerance: Double) {
+    when {
+      this < a -> this.shouldThrowMinimumExceptionOnBetween(a, b, tolerance)
+      this > b -> this.shouldThrowMaximumExceptionOnBetween(a, b, tolerance)
+      else     -> throw IllegalStateException()
+    }
+  }
+  
+  private fun Double.shouldThrowMinimumExceptionOnBetween(a: Double, b: Double, tolerance: Double) {
+    val message = "$this should be bigger than $a"
+    shouldThrowExceptionOnBetween(a, b, tolerance, message)
+  }
+  
+  private fun Double.shouldThrowMaximumExceptionOnBetween(a: Double, b: Double, tolerance: Double) {
+    val message = "$this should be smaller than $b"
+    shouldThrowExceptionOnBetween(a, b, tolerance, message)
+  }
+  
+  
+  private fun Double.shouldThrowExceptionOnBetween(a: Double,
+                                                   b: Double,
+                                                   tolerance: Double,
+                                                   message: String = "$this should be smaller than $b and bigger than $a"
+  ) {
+    shouldThrowAssertionError(message,
+                              { this.shouldBeBetween(a, b, tolerance) },
+                              { this shouldBe between(a, b, tolerance) })
+  }
+  
+  private fun Double.shouldThrowExceptionOnNotBetween(a: Double,
+                                                      b: Double,
+                                                      tolerance: Double,
+                                                      message: String = "$this should not be smaller than $b and should not be bigger than $a"
+  ) {
+    
+    shouldThrowAssertionError(message,
+                              { this.shouldNotBeBetween(a, b, tolerance) },
+                              { this shouldNotBe between(a, b, tolerance) })
+  }
+  
+  private fun onDoubleRange(block: (double: Double, valueToVariate: Double) -> Unit) {
+    (0.1..1.0 step 0.2).forEach { valueToVariate ->
+      assertAll(300, nonMaxValueDoubles) {
+        block(it, valueToVariate)
+      }
+    }
+  }
+  
+  private infix fun ClosedRange<Double>.step(step: Double): Iterable<Double> {
+    val sequence = generateSequence(start) { previous ->
+      if (previous == Double.POSITIVE_INFINITY) return@generateSequence null
```

Comment:
I think return@ is a code smell. It's more functional to have the return be the last value all the time. I'd simply change this to have the next two lines inside an else {}